sycl/source/device_selector.cpp

```diff
@@ -15,76 +15,133 @@
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
+
+// Utility function to check if device is of the preferred SYCL_BE.
+static bool isDeviceOfPreferredSyclBe(const device &Device) {
+  // Taking the version information from the platform gives us more useful
+  // information than the driver_version of the device.
+  const platform Platform = Device.get_info<info::device::platform>();
+  const std::string PlatformVersion =
+      Platform.get_info<info::platform::version>();
+
+  detail::pi::Backend PreferredBE = detail::pi::getPreferredBE();
+
+  if (PreferredBE == RT::Backend::SYCL_BE_PI_OPENCL) {
+    if (PlatformVersion.find("OpenCL") != std::string::npos) {
+      return true;
+    }
+  } else if (PreferredBE == RT::Backend::SYCL_BE_PI_CUDA) {
+    if (PlatformVersion.find("CUDA") != std::string::npos) {
+      return true;
+    }
+  }
+  return false;
+}
+
 device device_selector::select_device() const {
   vector_class<device> devices = device::get_devices();
   int score = -1;
   const device *res = nullptr;
-  for (const auto &dev : devices)
-    if (score < operator()(dev)) {
+  for (const auto &dev : devices) {
+    int dev_score = operator()(dev);
+    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL)) {
+      std::cout << "SYCL_PI_TRACE[1]: select_device(): -> score = " << score
+                << std::endl
+                << "SYCL_PI_TRACE[1]:   platform: "
+                << dev.get_info<info::device::platform>()
+                       .get_info<info::platform::version>()
+                << std::endl
+                << "SYCL_PI_TRACE[1]:   device: "
+                << dev.get_info<info::device::name>() << std::endl;
+    }
+
+    if (score < dev_score) {
       res = &dev;
-      score = operator()(dev);
+      score = dev_score;
     }
+  }
 
-  if (res != nullptr)
+  if (res != nullptr) {
+    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_BASIC)) {
+      std::cout << "SYCL_PI_TRACE[1]: select_device(): ->" << std::endl
```

Comment:
This message should somehow say that a specific device is selected.

Reply:
"->" is meant to mean that, but I agree we could be more specific

Reply:
Fixed.